File: py/generate-parentheses.py
class Solution:
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        dp = [[] for i in range(n+1)]
        dp[0].append("")
        
        for i in range(1,n+1):
            for j in range(i):
                inStrs = dp[j]
                outStrs = dp[i-j-1]
                for inStr in inStrs:
                    for outStr in outStrs:
                        newStr = "(" + inStr + ")" + outStr
                        dp[i].append(newStr)
        
        return dp[n]

File: cpp/generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<vector<string>> dp;

        vector<string> a;
        for (int i=0;i<=n;i++) dp.push_back(a);
        
        dp[0].push_back("");
        for (int i=1;i<=n;i++) {
            for (int j=0;j<i;j++) {
                vector<string> inStrs = dp[j];
                vector<string> outStrs = dp[i-j-1];
                for (string inStr: inStrs) {
                    for (string outStr: outStrs) {
                        string newStr = "(" + inStr + ")" + outStr;
                        dp[i].push_back(newStr);
                    }
                }
            }    
        }
        
        return dp[n];
    }
};